source/common/grpc/buffered_async_client.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/container/btree_map.h"
+
+namespace Envoy {
+namespace Grpc {
+
+enum class BufferState { Buffered, PendingFlush };
+
+// This class wraps bidirectional gRPC and provides message arrival guarantee.
+// It stores messages to be sent or in the process of being sent in a buffer,
+// and can track the status of the message based on the ID assigned to each message.
+// If a message fails to be sent, it can be re-buffered to guarantee its arrival.
+template <class RequestType, class ResponseType> class BufferedAsyncClient {
+public:
+  BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
+                      Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+      : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
+        client_(client) {}
+
+  ~BufferedAsyncClient() {
+    if (active_stream_ != nullptr) {
+      active_stream_ = nullptr;
+    }
+  }
+
+  void bufferMessage(RequestType& message) {
+    const auto buffer_size = message.ByteSizeLong();
+    if (current_buffer_bytes_ + buffer_size > max_buffer_bytes_) {
+      return;
+    }
+
+    auto id = publishId();
+    message_buffer_[id] = std::make_pair(BufferState::Buffered, message);
```

Comment:
In #17486 this id is set on the RequestType (CriticalAccessLogsMessage). In conjunction with the previous comment, maybe we should return `absl::optional<uint64_t>` with the id if buffered or null_opt if the buffer's full? The caller can then add the id to the message (although that kind of change via the message reference feels a little weird). Another possibility is to require that the id is passed obtained via publishId, used to call `set_id` on the message, and then passed to bufferMessage as a separate parameter.

Reply:
I missed it. I agree with to return `absl::optional<uint64_t>` them the caller set it to the message.